src/components/TextInput.tsx

```diff
@@ -0,0 +1,204 @@
+
+import { Colors, Indents, TextSize } from '@constants/styles';
+
+const minimumTouchableSize = 40;
+
+interface InputProps {
+    value: string;
+    required?: boolean;
+    keyboardType?: KeyboardTypeOptions;
+    autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters';
+    isEditable?: boolean;
+    multiline?: boolean;
+    maxLength?: number;
+    onChange: (value: string) => void;
+    errorHandler?: (value: boolean) => void;
+    label?: string;
+    placeholder?: string;
+    isPassword?: boolean;
+    validation?: { rule: RegExp; message: string }[];
+    styles?: {
+        wrapper?: StyleProp<ViewStyle>;
+        input?: StyleProp<TextStyle>;
+        label?: StyleProp<TextStyle>;
+        error?: StyleProp<TextStyle>;
+    };
+}
+
+const TextInput: React.FC<InputProps> = (props) => {
+    const {
+        required = false,
+        value: initialValue,
+        keyboardType,
+        autoCapitalize = 'none',
+        isEditable = true,
+        multiline = false,
+        maxLength,
+        placeholder,
+        errorHandler = () => {},
+        isPassword,
+        validation,
+        onChange,
+        styles,
+        label,
+    } = props;
+
+    const [value, setValue] = useState(initialValue);
+    const [error, setError] = useState('');
+    const [editable, setEditable] = useState(isEditable);
+    const [isFocused, setFocus] = useState(false);
+    const [secureText, setSecureText] = useState(isPassword);
+
+    useEffect(() => {
+        if (required && !value) {
+            errorHandler(true);
+        }
+    }, []);
+
+    // TODO: ###CF### Will produce extra re-render? Check and create custom useDidUpdate to get rid of it.
+    useEffect(() => {
+        setValue(initialValue);
+    }, [initialValue]);
+
+    useEffect(() => {
+        setEditable(isEditable);
+    }, [isEditable]);
+
+    const handlePasswordVisibilityChange = () => {
+        setSecureText(!secureText);
+    };
+
+    const handleChangeText = (value: string) => {
+        setValue(value);
+        if (validation?.length) {
+            const error = validation.find((el) => !el.rule.test(value));
+            setError(error ? error.message : '');
+            errorHandler(error ? true : required && !value);
+        } else {
+            errorHandler(required && !value);
+        }
+        onChange(value);
+    };
+
+    const handleFocus = () => {
+        setFocus(true);
+    };
+
+    const handleBlur = () => {
+        setFocus(false);
+        Keyboard.dismiss;
+    };
+
+    const errorCommonStyle: StyleProp<TextStyle> = [
+        generalStyles.error,
+        styles?.error,
+    ];
+
+    const labelCommonStyle: StyleProp<TextStyle> = [
+        generalStyles.label,
+        isFocused && generalStyles.labelFocused,
+    ];
+
+    const inputCommonStyle: StyleProp<TextStyle> = [
+        generalStyles.input,
+        isFocused && generalStyles.inputFocused,
+        !!error && generalStyles.inputError,
+    ];
```

Comment:
It's a little bit hard to focus on the component logic because of this styles. Why is it not in render?